File: algorithm/java/src/main/java/com/liyongquan/dp/CountSubstrings.java
package com.liyongquan.dp;

/**
 * 647. 回文子串
 * <p>
 * <p>
 * 给定一个字符串，你的任务是计算这个字符串中有多少个回文子串。
 * <p>
 * 具有不同开始位置或结束位置的子串，即使是由相同的字符组成，也会被视作不同的子串。
 * <p>
 *  
 * <p>
 * 示例 1：
 * <p>
 * 输入："abc"
 * 输出：3
 * 解释：三个回文子串: "a", "b", "c"
 * 示例 2：
 * <p>
 * 输入："aaa"
 * 输出：6
 * 解释：6个回文子串: "a", "a", "a", "aa", "aa", "aaa"
 *  
 * <p>
 * 提示：
 * <p>
 * 输入的字符串长度不会超过 1000 。
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/palindromic-substrings
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
public class CountSubstrings {
    /**
     * 先尝试暴力解法
     *
     * @param s
     * @return
     */
    public int countSubstrings(String s) {
        int len = s.length();
        if (len == 0) {
            return 0;
        }
        int res = len;
        for (int i = 2; i <= len; i++) {
            for (int j = 0; j <= len - i; j++) {
                if (isPalindromic(s.substring(j, j + i))) {
                    res++;
                }
            }
        }
        return res;
    }

    private boolean isPalindromic(String s) {
        int l = 0, r = s.length() - 1;
        while (l < r) {
            if (s.charAt(l++) != s.charAt(r--)) {
                return false;
            }
        }
        return true;
    }

    /**
     * 我们考虑如何降低回文串的判断效率
     * <p>
     * 假设f(i,j)为当前字符是否回文串，i<=j
     * <p>
     * f(i,j)=f(i+1,j-1)&&s[i]==s[j]
     * <p>
     * 那么我们可以使用dp来降低回文串的计算效率
     * <p>
     * 需要注意的是由于dp迭代是基于左下方的点决定的，因此我们需要从下往上迭代
     *
     * @param s
     * @return
     */
    public int countSubstrings2(String s) {
        int len = s.length();
        if (len == 0) {
            return 0;
        }
        boolean[][] dp = new boolean[len][len];
        int res = 0;
        for (int i = len - 1; i >= 0; i--) {
            for (int j = i; j < len; j++) {
                if (s.charAt(i) == s.charAt(j) && (i + 1 > j - 1 || dp[i + 1][j - 1])) {
                    dp[i][j] = true;
                    res++;
                }
            }
        }
        return res;
    }
}

File: algorithm/golang/array/countSubstrings.go
package array

// 给你两个字符串 s 和 t ，请你找出 s 中的非空子串的数目，这些子串满足替换 一个不同字符 以后，是 t 串的子串。换言之，请你找到 s 和 t 串中 恰好 只有一个字符不同的子字符串对的数目。
//
//比方说， "computer" and "computation" 只有一个字符不同： 'e'/'a' ，所以这一对子字符串会给答案加 1 。
//
//请你返回满足上述条件的不同子字符串对数目。
//
//一个 子字符串 是一个字符串中连续的字符。
//
//
//
//示例 1：
//
//输入：s = "aba", t = "baba"
//输出：6
//解释：以下为只相差 1 个字符的 s 和 t 串的子字符串对：
//("aba", "baba")
//("aba", "baba")
//("aba", "baba")
//("aba", "baba")
//("aba", "baba")
//("aba", "baba")
//加粗部分分别表示 s 和 t 串选出来的子字符串。
//示例 2：
//输入：s = "ab", t = "bb"
//输出：3
//解释：以下为只相差 1 个字符的 s 和 t 串的子字符串对：
//("ab", "bb")
//("ab", "bb")
//("ab", "bb")
//加粗部分分别表示 s 和 t 串选出来的子字符串。
//示例 3：
//输入：s = "a", t = "a"
//输出：0
//示例 4：
//
//输入：s = "abe", t = "bbc"
//输出：10
//
//
//提示：
//
//1 <= s.length, t.length <= 100
//s 和 t 都只包含小写英文字母。
//
//来源：力扣（LeetCode）
//链接：https://leetcode.cn/problems/count-substrings-that-differ-by-one-character
//著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

// 暴力解法。时间复杂度O(n^4)，勉强通过
func countSubstrings(s string, t string) int {
	l1, l2 := len(s), len(t)
	mxLen := l1
	if l2 < l1 {
		mxLen = l2
	}
	check := func(i, j, k int) bool {
		cnt := 0
		for l := 0; l < k; l++ {
			if s[i+l] != t[j+l] {
				cnt++
			}
			if cnt > 2 {
				return false
			}
		}
		return cnt == 1
	}
	res := 0
	for k := 1; k <= mxLen; k++ {
		for i := 0; i <= l1-k; i++ {
			for j := 0; j <= l2-k; j++ {
				if check(i, j, k) {
					res++
				}
			}
		}
	}
	return res
}

// 优化解法
func countSubstrings2(s string, t string) int {
	res := 0
	for i := 0; i < len(s); i++ {
		for j := 0; j < len(t); j++ {
			diff := 0
			for k := 0; i+k < len(s) && j+k < len(t); k++ {
				if s[i+k] != s[j+k] {
					diff++
				}
				if diff == 1 {
					res++
				} else if diff > 1 {
					break
				}
			}
		}
	}
	return res
}
